### PR TITLE
test(e2e, playwright): email-sign-up

### PR DESCRIPTION
**Pull request by kensiecodes**
Checklist:

- [ x ] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [ x ] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [ x ] My pull request targets the `main` branch of freeCodeCamp.
- [ x ] I have tested these changes either locally on my machine, or GitPod.

Ref #51705 

All tests pass locally!

Submitted this commit with a --no-verify flag due to a bizarre memory issue when the commit attempted to lint the two files. Still really don't know why this is happening, but seems to be a somewhat common problem from Google research. It generates a large error, but mainly "FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory" sticks out, and after allocating more memory to JS/Node via Terminal it still occurs.

### CODE REVIEW

**Comment by kensiecodes** [comment]
Oh, and camperbot can't code >:(

File: e2e/email-sign-up.spec.ts
Comment (naomi-lgbt) [inline e2e/email-sign-up.spec.ts:13]:
```suggestion
test.beforeEach(async ({ page }) => {
  await page.goto('/email-sign-up');
});
```

We should ideally run the tests in isolation.

File: e2e/email-sign-up.spec.ts
Comment (naomi-lgbt) [inline e2e/email-sign-up.spec.ts:4]:
```suggestion
import { test, expect } from '@playwright/test';
```

Remove the unused translations import and global `page` variable.

File: e2e/email-sign-up.spec.ts
Comment (naomi-lgbt) [inline e2e/email-sign-up.spec.ts:17]:
```suggestion
  test('The page renders with correct title', async ({ page }) => {
```

Because we can get `page` directly from the function args.

File: e2e/email-sign-up.spec.ts
Comment (naomi-lgbt) [inline e2e/email-sign-up.spec.ts:23]:
```suggestion
  test('The page renders the correct content', async ({ page }) => {
```

File: e2e/email-sign-up.spec.ts
Comment (naomi-lgbt) [inline e2e/email-sign-up.spec.ts:31]:
```suggestion
  test('The sign up button renders and functions', async({ page }) => {
```

File: e2e/email-sign-up.spec.ts
Comment (naomi-lgbt) [inline e2e/email-sign-up.spec.ts:27]:
```suggestion
        page.getByText(
          'freeCodeCamp is a proven path to your first software developer job.
        )
```

We aren't using a custom locator here, so we should use the `getByText` method instead.

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]
This will also need the code formatter to be run.

**Comment by kensiecodes** [comment]
> This will also need the code formatter to be run.

@naomi-lgbt Do you have any suggestions for the memory issue that isn't allowing me to pass it through the formatter? I'm not sure how else to commit or bypass and I'm not even sure the source of the error.

File: client/src/pages/email-sign-up.tsx
Comment (naomi-lgbt) [inline client/src/pages/email-sign-up.tsx:64]:
```suggestion
      <Row data-playwright-test-label='user-email-buttons'>
```

File: client/src/pages/email-sign-up.tsx
Comment (naomi-lgbt) [inline client/src/pages/email-sign-up.tsx:101]:
```suggestion
          data-playwright-test-label='email-sign-up-button'
```

File: client/src/pages/email-sign-up.tsx
Comment (naomi-lgbt) [inline client/src/pages/email-sign-up.tsx:157]:
```suggestion
            <IntroDescription data-playwright-test-label='learn-read-this-section' />
```

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]
Here's a couple.

You can also try `pnpm run format:prettier` locally to see if it doesn't run out of memory outside of the commit hook.

**Comment by kensiecodes** [comment]
Not sure why the issue was occurring but seems to have resolved itself, I successfully committed with formatting! Also merged all your changes @naomi-lgbt thank you so much for your review. Fingers crossed it's all good now but just let me know.

File: e2e/email-sign-up.spec.ts
Comment (naomi-lgbt) [inline e2e/email-sign-up.spec.ts:28]:
```suggestion
  test('should render correctly', async ({ page }) => {
    await expect(page).toHaveTitle(`Email Sign Up | freeCodeCamp.org`);
    await expect(
      page.getByText(
        'freeCodeCamp is a proven path to your first software developer job.'
      )
    ).toBeVisible();
    const button = page.getByTestId('email-sign-up-button');
    await expect(button).toBeVisible();
    await button.click();
    await expect(page).toHaveURL(/.*\/learn\/?$/);
  });
```
We can probably combine all of these into a single rendering test.

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

**Comment by kensiecodes** [comment]
Hey @naomi-lgbt just poking for review! :)

**Comment by moT01** [comment]
Looks like some tests are failing here @kensiecodes - I think you need to rebase. Some things were broken and have since been fixed. A rebase should fix most of the tests. Also, it looks like you need to update the snapshot.

**Comment by kensiecodes** [comment]
@huyenltnguyen Thank you so much! That should have clicked for me as being a simple fix lol. Sorry, still getting used to working on a massive repo. Will push fix today!

**Comment by raisedadead** [comment]
No worries, @kensiecodes we all start somewhere, and we appreciate your help with these changes. Here are some steps that might be of help:

1. You would want to rebase the repo from the `upstream`. This can be done in two ways:

   a) Either do a "rebase" of the main branch:

      ```
      git checkout test/e2e/emailsignup
      git pull --rebase upstream main
      ```
      
      This will add your commits on top of the commits from the main branch.

   OR

   b) Merge the main branch onto your branch:

      ```
      git checkout test/e2e/emailsignup
      git fetch upstream
      git merge upstream/main
      ```

      This will add a merge commit from the main branch.

   You can just pick one of these methods.

2. Once you have rebased the main branch, you can update the snapshots for some related unit tests since they need to be updated now.

   ```
   pnpm install

   # This will show you the failing test in the client
   pnpm run test-client
   ```

   Update using (this is a quick way):

   ```
   npx jest client -u
   ```

Update your changes to the PR and we are all set. If you need any help, please reach out.

**Review by huyenltnguyen — COMMENTED** [review]
Hey @kensiecodes, thank you for your contribution and your patience!

I went ahead and updated your PR to shorten the review loop a little. The changes can be seen here: https://github.com/freeCodeCamp/freeCodeCamp/commit/0b25b24b24d257680e9381b0d5e6952acea6235d. I also merged the changes in #52019 and added some improvements, which is why the commit has two authors.

---
What I changed:
- Removed the test IDs and wrote tests using the elements' role and text
- Updated the test suite to cover the following scenarios:
  - When the user goes to the page but is not signed in
  - When the user is signed in, goes to the page and clicks No
  - When the user is signed in, goes to the page and clicks Yes

Test result:

Screenshot 2023-11-09 at 00 02 25

---

To maintainers: I made quite a lot of changes so I'm not confident approving my own code. We will need another set of eyes here 😅 And if there is anything looks weird to you, that's all mine.

**Review by raisedadead — APPROVED** [review]

**Review by naomi-lgbt — APPROVED** [review]

**Pull request merged**
